nautobot/core/views/renderers.py

```diff
@@ -0,0 +1,217 @@
+                else:
+                    pk_list = request.POST.getlist("pk")
+                    initial = {
+                        "pk": pk_list,
+                        "return_url": return_url,
+                    }
+                    form = form_class(initial=initial)
+                table = self.construct_table(view, pk_list=pk_list)
+            elif view.action == "bulk_create":
+                form = view.get_form()
+                if request.data:
+                    table = data.get("table")
+            elif view.action == "bulk_update":
+                if request.POST.get("_all") and view.filterset_class is not None:
+                    pk_list = [obj.pk for obj in view.filterset_class(request.POST, view.queryset.only("pk")).qs]
+                else:
+                    pk_list = request.POST.getlist("pk")
+
+                if "_apply" in request.POST:
+                    form = form_class(model, request.POST)
+                else:
+                    # Include the PK list as initial data for the form
+                    initial_data = {"pk": pk_list}
+                    form = form_class(model, initial=initial_data)
+
+                restrict_form_fields(form, request.user)
+                table = self.construct_table(view, pk_list=pk_list)
+
+        context = {
+            "action_buttons": valid_actions,
+            "changelog_url": changelog_url,
+            "content_type": content_type,
+            "editing": instance.present_in_database,
+            "filter_form": self.get_filter_form(view, request),
+            "form": form,
+            "object": instance,
+            "obj": instance,  # NOTE: This context key is deprecated in favor of `object`.
+            "obj_type": view.queryset.model._meta.verbose_name,  # NOTE: This context key is deprecated in favor of `verbose_name`.
+            "obj_type_plural": view.queryset.model._meta.verbose_name_plural,  # NOTE: This context key is deprecated in favor of `verbose_name_plural`.
+            "permissions": permissions,
+            "return_url": return_url,
+            "table": table if table is not None else data.get("table", None),
+            "table_config_form": TableConfigForm(table=table) if table else None,
+            "verbose_name": view.queryset.model._meta.verbose_name,
+            "verbose_name_plural": view.queryset.model._meta.verbose_name_plural,
+        }
+        if view.action == "retrieve":
+            context.update(view.get_extra_context(request, instance))
+        else:
+            if view.action == "bulk_create":
+                context.update(
+                    {
+                        "active_tab": "csv-data",
+                        "fields": view.bulk_create_form_class(model).fields if view.bulk_create_form_class else None,
+                    }
+                )
+            context.update(view.get_extra_context(request, instance=None))
```

Comment:
Feels weird that `get_extra_context` is *only* called for `retrieve` and `bulk_create` actions - at the very least the docstring for this method should make it really clear that it doesn't apply to any other actions, but maybe we should always call it?

Reply:
It does get called for other actions no? Line 203?